apps/voting/contracts/Voting.sol

```diff
@@ -291,7 +277,9 @@ contract Voting is IForwarder, AragonApp {
     * @return True if the given vote can be executed, false otherwise
     */
     function canExecute(uint256 _voteId) public view voteExists(_voteId) returns (bool) {
-        return _canExecute(votes[_voteId]);
+        Vote storage vote_ = votes[_voteId];
+        bool can = _canExecute(vote_);
+        return vote_.disputableStatus == DisputableStatus.Invalid ? can : can && _canProceed(vote_.actionId);
```

Comment:
I don't fully understand the Agreements app, but what's the reason for checking `_canProceed` here? Does the `_isPausedOrCancelled` function called by `_canExecute` not cover this?

Reply:
Yes, that sounds right. Actually see this comment: https://github.com/aragon/aragon-apps/pull/1149#discussion_r430596773

Reply:
> what's the reason for checking `_canProceed` here?

The rationale is that some of the things happening in the Agreement app are being cached in the side of the Voting app. This is the case of a vote being challenged (paused) for example. 

Note that if we don't cache this status there is no security issue tho, I think it's better to do it to ensure a better UX. For example, if you try to execute a vote being challenged, therefore you will try to close the Agreement action and it will fail. Having these caches allows us to improve the querying methods, otherwise `canVote` or `canExecute` could return a false positive. 

> Does the `_isPausedOrCancelled` function called by `_canExecute` not cover this?

Potentially, we cannot be 100% sure our cache works totally perfect. We can now, but the Agreement app could be upgraded or whatever. Thus, I think it's a good idea to properly check against the Agreement app if an action can proceed in these query methods.

What do you guys think?

Reply:
Ah yeah, I understand now. It seems reasonable. However, would like to point out, in the case an agreement action is not ready `vote()` is only restricted by agreements if it can also execute the vote. Where as `canVote()` is restricted by agreements whether it can execute it or not. So I think there could be situations where `canVote()` fails but `vote()` is still executable.

Reply:
It sounds good to me, although a little bit inconsistent, as for casting a vote we rely on the local cache only, instead of also going to Agreements to check, right?

Reply:
> Ah yeah, I understand now. It seems reasonable. However, would like to point out, in the case an agreement action is not ready `vote()` is only restricted by agreements if it can also execute the vote. Where as `canVote()` is restricted by agreements whether it can execute it or not. So I think there could be situations where `canVote()` fails but `vote()` is still executable.

Oh, I didn’t have Github page refreshed while I was trying to analyze this, so I didn’t see your previous response, but yes, I mean essentially the same (you explained it better actually :wink: )

Reply:
That totally makes sense, added the `_canProceed` check to the internal methods for now then